sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -66,22 +66,36 @@ trait Column extends DataFrame {
    */
   def isComputable: Boolean
 
+  /** Removes the top project so we can get to the underlying plan. */
+  private def stripProject(p: LogicalPlan): LogicalPlan = p match {
+    case Project(_, child) => child
+    case p => sys.error("Unexpected logical plan (expected Project): " + p)
+  }
+
   private def computableCol(baseCol: ComputableColumn, expr: Expression) = {
     val plan = Project(Seq(expr match {
       case named: NamedExpression => named
       case unnamed: Expression => Alias(unnamed, "col")()
-    }), baseCol.plan)
+    }), stripProject(baseCol.plan))
```

Comment:
Nit: I would consider splitting this up since matching in the constructor is not super easy to read.